ports/stm32/mboot/main.c

```diff
@@ -1030,21 +1086,23 @@ static int dfu_handle_tx(int cmd, int arg, int len, uint8_t *buf, int max_len) {
             case DFU_CMD_NONE:
                 break;
             case DFU_CMD_EXIT:
-                dfu_state.status = DFU_STATUS_MANIFEST;
+                dfu_state.state = DFU_STATE_MANIFEST;
                 break;
             case DFU_CMD_UPLOAD:
-                dfu_state.status = DFU_STATUS_UPLOAD_IDLE;
+                dfu_state.state = DFU_STATE_UPLOAD_IDLE;
                 break;
             case DFU_CMD_DNLOAD:
-                dfu_state.status = DFU_STATUS_BUSY;
+                dfu_state.state = DFU_STATE_BUSY;
                 break;
+            default:
+                dfu_state.state = DFU_STATE_BUSY;
         }
-        buf[0] = 0;
-        buf[1] = dfu_state.cmd; // TODO is this correct?
-        buf[2] = 0;
-        buf[3] = 0;
-        buf[4] = dfu_state.status;
-        buf[5] = 0;
+        buf[0] = dfu_state.status; // bStatus
+        buf[1] = 0;  // bwPollTimeout (ms)
+        buf[2] = 0;  // bwPollTimeout (ms)
+        buf[3] = 0;  // bwPollTimeout (ms)
+        buf[4] = dfu_state.state; // bState
+        buf[5] = dfu_state.error;  // iString
```

Comment:
When I originally wrote this I knew I didn't have this fully correct, even though it all worked.  So this is a good change; I assume it's been tested?

Reply:
yes these all appear to line up better with the dfu spec, and do/still work with dfu-util as well as pydfu.
I've also extended the changes further in https://github.com/micropython/micropython/pull/5730 with further usage and testing.